### PR TITLE
Pass proxy environment variables to the building container when building node images

### PR DESCRIPTION
**Pull request by Zumium**
When used in a restricted network environment, KIND may not be able to download necessary resources directly from the Internet during the building process, which makes the building fail. Adding` --http-proxy` option to the "build node-image" command helps to solve the problem by setting up the building container with `http_proxy` and `https_proxy` environment variables and making the downloading to be able to use the provided proxy.

### CODE REVIEW

**Comment by aojea** [comment]
we inherit proxy variables from the environment, is this not the case with building ?

https://kind.sigs.k8s.io/docs/user/quick-start/#configure-kind-to-use-a-proxy

I don't expect to add flags for this

File: pkg/build/nodeimage/buildcontext.go [hunk 1]
```diff
@@ -135,6 +136,9 @@ func (c *buildContext) buildImage(bits kube.Bits) error {
 		"docker", "commit",
 		// we need to put this back after changing it when running the image
 		"--change", `ENTRYPOINT [ "/usr/local/bin/entrypoint", "/sbin/init" ]`,
+		// remove http proxy setting
```
Comment (aojea) [inline pkg/build/nodeimage/buildcontext.go:139]: why remove?

File: pkg/build/nodeimage/buildcontext.go [hunk 2]
```diff
@@ -314,16 +318,23 @@ func (c *buildContext) createBuildContainer() (id string, err error) {
...
+	runArgs := []string{
+		"-d", // make the client exit while the container continues to run
+		// the container should hang forever, so we can exec in it
+		"--entrypoint=sleep",
+		"--name=" + id,
+		"--platform=" + dockerBuildOsAndArch(c.arch),
+		"--security-opt", "seccomp=unconfined", // ignore seccomp
+		"--network=host",
+	}
+	if c.httpProxy != "" {
```
Comment (aojea) [inline pkg/build/nodeimage/buildcontext.go:330]: this can be obtained from https://github.com/kubernetes-sigs/kind/blob/40c81f187425254daf2bf84360a6257a278252df/pkg/cluster/internal/providers/common/proxy.go#L35-L37

File: pkg/cmd/kind/build/nodeimage/nodeimage.go [hunk 3]
```diff
@@ -86,6 +87,12 @@ func NewCommand(logger log.Logger, streams cmd.IOStreams) *cobra.Command {
 		"",
 		"architecture to build for, defaults to the host architecture",
 	)
+	cmd.Flags().StringVar(
```
Comment (aojea) [inline pkg/cmd/kind/build/nodeimage/nodeimage.go:90]: no flags

File: pkg/build/nodeimage/buildcontext.go (hunk 1 above)
Comment (Zumium) [reply pkg/build/nodeimage/buildcontext.go:139]: It's because during my testing, I found that the control panel wasn't able to start up with my built node image, and the built image was carrying the proxy settings. When I removed them, it worked.

**Comment by Zumium** [comment]
@aojea Hi, aojea.  The "http_proxy" and "https_proxy" env vars are not working with building yet because they are not passed to the building container, which is what happened when I tested it. If no flags is the preferred way, I'll adjust my code to read proxy settings from env vars instead of a flag.

File: pkg/build/nodeimage/buildcontext.go (hunk 2 above)
Comment (Zumium) [reply pkg/build/nodeimage/buildcontext.go:330]: It seems that this `GetProxyEnvs` function is for using in provisioning clusters, because it requires a cluster config as its parameter, which we don't have when building node images. I'll just read from env vars directly.

File: pkg/build/nodeimage/buildcontext.go [hunk 4]
```diff
@@ -314,16 +326,29 @@ func (c *buildContext) createBuildContainer() (id string, err error) {
...
 	random := rand.New(rand.NewSource(time.Now().UnixNano())).Int31()
 	id = fmt.Sprintf("kind-build-%d-%d", time.Now().UTC().Unix(), random)
+	runArgs := []string{
+		"-d", // make the client exit while the container continues to run
+		// the container should hang forever, so we can exec in it
+		"--entrypoint=sleep",
+		"--name=" + id,
+		"--platform=" + dockerBuildOsAndArch(c.arch),
+		"--security-opt", "seccomp=unconfined", // ignore seccomp
+		"--network=host",
```
Comment (aojea) [inline pkg/build/nodeimage/buildcontext.go:336]: this was not here before, was it?

File: pkg/build/nodeimage/buildcontext.go [hunk 5]
```diff
@@ -135,6 +144,9 @@ func (c *buildContext) buildImage(bits kube.Bits) error {
 		"docker", "commit",
 		// we need to put this back after changing it when running the image
 		"--change", `ENTRYPOINT [ "/usr/local/bin/entrypoint", "/sbin/init" ]`,
+		// remove proxy settings since they're for the building process
+		// and should not be carried with the built image
+		"--change", `ENV HTTP_PROXY="" HTTPS_PROXY="" NO_PROXY=""`,
```
Comment (aojea) [inline pkg/build/nodeimage/buildcontext.go:149]: I'm not familiar with this, what is this change required?

File: pkg/build/nodeimage/buildcontext.go (hunk 4 above)
Comment (Zumium) [reply pkg/build/nodeimage/buildcontext.go:336]: The line `--network=host` is added by me, the others were there before. The purpose of `--network=host` is to be able to use local host proxy like "http://127.0.0.1:1080".

File: pkg/build/nodeimage/buildcontext.go (hunk 5 above)
Comment (Zumium) [reply pkg/build/nodeimage/buildcontext.go:149]: Sorry I may not fully understand your question.  The `--change` option accepts a Dockerfile instruction and applies it to the outputted image. Detail is here: https://docs.docker.com/engine/reference/commandline/commit/#options

File: pkg/build/nodeimage/buildcontext.go (hunk 4 above)
Comment (aojea) [reply pkg/build/nodeimage/buildcontext.go:336]: you can not submit a PR with multiple changes without informing the reviewers, this is a change in behavior, it requires additional discussion, and in this case it seems that is only to fit your personal workflow

Comment (aojea) [reply pkg/build/nodeimage/buildcontext.go:336]: please do only one change at a time

File: pkg/build/nodeimage/buildcontext.go (hunk 5 above)
Comment (aojea) [reply pkg/build/nodeimage/buildcontext.go:149]: but why do we need to use it?

File: pkg/build/nodeimage/buildcontext.go (hunk 1 above)
Comment (aojea) [reply pkg/build/nodeimage/buildcontext.go]: this does not gives much confidence on the solution proposed :/

File: pkg/build/nodeimage/buildcontext.go (hunk 5 above)
Comment (Zumium) [reply pkg/build/nodeimage/buildcontext.go:149]: Right now it's only based on my local testing result (control panel won't start successfully if they're not removed), and I haven't figured out what's going on underneath. I'll offer a detailed explanation in a few days cause I can only do this after work, please bear with me.

Comment (aojea) [reply pkg/build/nodeimage/buildcontext.go:149]: there is no rush, take your time, but if we add any option we need to be sure we don't break any existing workflow

File: pkg/build/nodeimage/buildcontext.go (hunk 4 above)
Comment (liangyuanpeng) [reply pkg/build/nodeimage/buildcontext.go:336]: > The purpose of --network=host is to be able to use local host proxy like "http://127.0.0.1:1080/".

@Zumium 
This is not necessary for the current PR, you can still use the proxy even without the host network, doesn't it?

File: pkg/build/nodeimage/buildcontext.go (hunk 5 above)
Comment (BenTheElder) [reply pkg/build/nodeimage/buildcontext.go:149]: this is necessary so we don't wind up saving the env to the saved image

File: pkg/build/nodeimage/buildcontext.go (hunk 4 above)
Comment (BenTheElder) [reply pkg/build/nodeimage/buildcontext.go:336]: > The line --network=host is added by me, the others were there before. The purpose of --network=host is to be able to use local host proxy like "http://127.0.0.1:1080/".

We should drop this, because you can't run node containers this way anyhow and will need a different solution for proxy?

**Comment by BenTheElder** [comment]
sorry for the delay, we should still make one of the requested changes re: host network

File: pkg/build/nodeimage/buildcontext.go (hunk 4 above)
Comment (Zumium) [reply pkg/build/nodeimage/buildcontext.go:336]: @liangyuanpeng Well, I'm afraid it's necessary especially for a proxy running on localhost. Imagine your proxy server is running locally on host, for example http://127.0.0.1:1080, which I believe is a common case. If the address is directly set to the container and not using host network, "127.0.0.1" will refer to the container's localhost, and the programs inside the container will fail to connect to proxy because they're not connecting to the right place.

Comment (Zumium) [reply pkg/build/nodeimage/buildcontext.go:336]: @BenTheElder I believe that the core of the problem is the meaning of "localhost" changes inside & outside of the container, and proxies usually bind on localhost for safety reason, preventing a process inside container from connecting to it directly. If host network is not accepted, we may have to add some kind of relay-er, and it will be a more complex solution.

Comment (aojea) [reply pkg/build/nodeimage/buildcontext.go:336]: you can use the IP of your host, proxies use to listen in all host ip addresses, is easier to modify your proxy than force this pod to run with hostnetwork

Comment (Zumium) [reply pkg/build/nodeimage/buildcontext.go:336]: It's also ok to just keep KIND simple and leave the worry to users. I'll modify my code later.

Comment (aojea) [reply pkg/build/nodeimage/buildcontext.go:336]: I don't disagree, but this was this way for years and everytime we change something that sounds simple, it always have consequences

**Comment by aojea** [comment]
can you squash all commits?

**Comment by Zumium** [comment]
> can you squash all commits?

Done.

**Comment by Zumium** [comment]
Thank everyone for helping! What a coincidence that today happens to be 除夕(chu xi) festival, which marks the beginning of a new Chinese year. Again, thank you everyone and happy Chinese new year!

**Comment by liangyuanpeng** [comment]
happy new year :)
